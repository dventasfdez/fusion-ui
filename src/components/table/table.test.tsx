import React from 'react';
import Table, {TBody, TElement, THead, TRow, ITableProps} from './table';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import {prettyDOM, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const leftElements = [
  <div key={1} className="checkbox-container">
    <input
      id="checkbox"
      onChange={(e) => {
        return e;
      }}
      type="checkbox"
    />
  </div>,
];

const rightElements = [
  <div key={1} className="checkbox-container">
    <input
      id="checkbox"
      onChange={(e) => {
        return e;
      }}
      type="checkbox"
    />
  </div>,
];

const elements = [
  ['Content', 'Content', 'Content', 'Content', 'Content', 'Content', 'Content'],
  ['Content', 'Content', 'Content', 'Content', 'Content', 'Content', 'Content'],
  ['Content', 'Content', 'Content', 'Content', 'Content', 'Content', 'Content'],
];

const newElements = [
  ['Content2', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4'],
  ['Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4'],
  ['Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4', 'Content 2 3 4'],
];

const headElements = ['Label', 'Label', 'Label', 'Label', 'Label', 'Label', 'Label'];

const TableTest = (args: ITableProps) => {
  return (
    <div className="tag-ds App" style={{height: '30vh', width: '50vw'}}>
      <Table {...args}>
        <THead>
          <TRow>
            {leftElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {headElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {rightElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
          </TRow>
        </THead>
        <TBody data-testid="table">
          {args.content.map((_, idx) => {
            if (args.tr)
              return (
                <tr key={idx} onClick={args.onClick}>
                  {leftElements.map((elem, idk) => (
                    <TElement key={'test' + idk}>{elem}</TElement>
                  ))}
                  {_.map((elem, idk) => (
                    <TElement key={idx + 'string' + idk}>{elem}</TElement>
                  ))}
                  {rightElements.map((elem, idk) => (
                    <TElement key={'test' + idk}>{elem}</TElement>
                  ))}
                </tr>
              );
            if (args.unvalid) {
              return (
                <div key={idx} onClick={args.onClick}>
                  {leftElements.map((elem, idk) => (
                    <TElement key={'test' + idk}>{elem}</TElement>
                  ))}
                  {_.map((elem, idk) => (
                    <TElement key={idx + 'string' + idk}>{elem}</TElement>
                  ))}
                  {rightElements.map((elem, idk) => (
                    <TElement key={'test' + idk}>{elem}</TElement>
                  ))}
                </div>
              );
            }
            return (
              <TRow key={idx} onClick={args.onClick}>
                {leftElements.map((elem, idk) => (
                  <TElement key={'test' + idk}>{elem}</TElement>
                ))}
                {_.map((elem, idk) => (
                  <TElement key={idx + 'string' + idk}>{elem}</TElement>
                ))}
                {rightElements.map((elem, idk) => (
                  <TElement key={'test' + idk}>{elem}</TElement>
                ))}
              </TRow>
            );
          })}
        </TBody>
      </Table>
    </div>
  );
};

const InvalidTableTest = (args: ITableProps) => {
  return (
    <div className="tag-ds App" style={{height: '30vh', width: '50vw'}}>
      <Table {...args}>
        <THead>
          <TRow>
            {leftElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {headElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {rightElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
          </TRow>
        </THead>
        <TBody data-testid="table">
          {elements.map((_, idx) => (
            <div key={idx}>
              {leftElements.map((elem, idk) => (
                <TElement key={'test' + idk}>{elem}</TElement>
              ))}
              {_.map((elem, idk) => (
                <TElement key={idx + 'string' + idk}>{elem}</TElement>
              ))}
              {rightElements.map((elem, idk) => (
                <TElement key={'test' + idk}>{elem}</TElement>
              ))}
            </div>
          ))}
        </TBody>
      </Table>
    </div>
  );
};

const TheadTableTest = (args: ITableProps) => {
  return (
    <div className="tag-ds App" style={{height: '30vh', width: '50vw'}}>
      <Table {...args}>
        <THead>
          <TRow>
            {leftElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {headElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
            {rightElements.map((elem, idk) => (
              <TElement key={'test' + idk}>{elem}</TElement>
            ))}
          </TRow>
        </THead>
        <TBody data-testid="table">
          {elements.map((_, idx) => (
            <TRow key={idx} onClick={args.onClick}>
              {leftElements.map((elem, idk) => (
                <TElement key={'test' + idk}>{elem}</TElement>
              ))}
              {_.map((elem, idk) => (
                <TElement key={idx + 'string' + idk}>{elem}</TElement>
              ))}
              {rightElements.map((elem, idk) => (
                <TElement key={'test' + idk}>{elem}</TElement>
              ))}
            </TRow>
          ))}
        </TBody>
      </Table>
    </div>
  );
};

test('renders and matches snapshot', () => {
  const component = renderer.create(<TableTest content={elements} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders and click on row', () => {
  const {getByTestId} = render(<TableTest content={elements} data-testid="table" />);
  const row = getByTestId('table-body-row-1');
  userEvent.click(row);
  expect(row).toHaveClass('active');
});

test('render disabled table', () => {
  const {getByTestId} = render(<TableTest content={elements} disabled={true} data-testid="table" />);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});

test('render with invalid body', () => {
  const {getByTestId} = render(<InvalidTableTest disabled={true} data-testid="table" />);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});

test('render with invalid thead', () => {
  const {getByTestId} = render(<TheadTableTest disabled={true} data-testid="table" />);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});

test('render and change active row', () => {
  const {getByTestId} = render(<TableTest content={elements} data-testid="table" />);
  const row = getByTestId('table-body-row-1');
  userEvent.click(row);
  expect(row).toHaveClass('active');
  userEvent.click(row);
  expect(row).not.toHaveClass('active');
});

test('rerender table with new data and fake click and real onClick', () => {
  const rOnClick = jest.fn();
  const {rerender, container, getByTestId} = render(<TableTest content={elements} data-testid="table" onClick={rOnClick} />);
  expect(container).toBeDefined();
  const row = getByTestId('table-body-row-1');
  userEvent.click(row);
  expect(rOnClick).toBeCalled();
  rerender(<TableTest unValid data-testid="table" content={newElements} />);
  rerender(<TableTest unValid data-testid="table" content={newElements} />);
  rerender(<TableTest unValid={false} data-testid="tables" content={newElements} onClick={rOnClick} />);
  rerender(<TableTest data-testid="tables" content={newElements} onClick={rOnClick} tr />);
  expect(container).toBeDefined();
});

test('render with tr', () => {
  const {getByTestId} = render(<TableTest content={elements} data-testid="table" tr />);
  const row = getByTestId('table-body-row-1').children;
  expect(row).toHaveLength(9);
});
