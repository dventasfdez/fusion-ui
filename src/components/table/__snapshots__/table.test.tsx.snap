// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<div
  className="fusion-ui App"
  style={
    Object {
      "height": "30vh",
      "width": "50vw",
    }
  }
>
  <div
    className="table-rotate-device"
  >
    <span
      className="material-icons"
    >
      screen_rotation
    </span>
    <span
      className="table-rotate-device-text"
    >
      Rotate your device
    </span>
  </div>
  <div
    className="table-wrapper"
  >
    <table
      className=""
      content={
        Array [
          Array [
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
          ],
          Array [
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
          ],
          Array [
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
            "Content",
          ],
        ]
      }
      data-testid=""
      role="none"
    >
      <thead
        className=""
      >
        <tr
          className=""
          tabIndex={0}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
      </thead>
      <tbody
        className="rest"
        data-testid="table"
      >
        <tr
          className=""
          data-testid="table-body-row-0"
          onClick={[Function]}
          tabIndex={0}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
        <tr
          className=""
          data-testid="table-body-row-1"
          onClick={[Function]}
          tabIndex={0}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
        <tr
          className=""
          data-testid="table-body-row-2"
          onClick={[Function]}
          tabIndex={0}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
